jcl/src/java.base/share/classes/java/lang/invoke/MethodHandle.java

```diff
@@ -1246,57 +1124,66 @@ private static final MethodHandle resolveInvokeDynamic(long j9class, String name
+				result = bsm.invoke(staticArgs[0], staticArgs[1], staticArgs[2], staticArgs[3]);
 				break;
 			case 5:
-				cs = (CallSite) bsm.invoke(staticArgs[0], staticArgs[1], staticArgs[2], staticArgs[3], staticArgs[4]);
+				result = bsm.invoke(staticArgs[0], staticArgs[1], staticArgs[2], staticArgs[3], staticArgs[4]);
 				break;
 			case 6:
-				cs = (CallSite) bsm.invoke(staticArgs[0], staticArgs[1], staticArgs[2], staticArgs[3], staticArgs[4], staticArgs[5]);
+				result = bsm.invoke(staticArgs[0], staticArgs[1], staticArgs[2], staticArgs[3], staticArgs[4], staticArgs[5]);
 				break;
 			case 7:
-				cs = (CallSite) bsm.invoke(staticArgs[0], staticArgs[1], staticArgs[2], staticArgs[3], staticArgs[4], staticArgs[5], staticArgs[6]);
+				result = bsm.invoke(staticArgs[0], staticArgs[1], staticArgs[2], staticArgs[3], staticArgs[4], staticArgs[5], staticArgs[6]);
 				break;
 			default:
-				cs = (java.lang.invoke.CallSite) bsm.invokeWithArguments(staticArgs);
+				result = bsm.invokeWithArguments(staticArgs);
 				break;
 			}
-			
-			if (cs != null) {
-				MethodType callsiteType = cs.type();
-				if (callsiteType != type) {
-					throw WrongMethodTypeException.newWrongMethodTypeException(type, callsiteType);
+
+			/* result validation */
+			if (indyNotCondy) { /* invoke dynamic */
+				if (result != null) {
+					MethodType callsiteType = ((CallSite)result).type();
+					if (callsiteType != (MethodType)type) {
+						throw WrongMethodTypeException.newWrongMethodTypeException((MethodType)type, callsiteType);
+					}
+					result = ((CallSite)result).dynamicInvoker();
 				}
-				result = cs.dynamicInvoker();
+			} else { /* constant dynamic */
+				result = MethodHandles.identity((Class<?>)type).invoke(result);
 			}
 		} catch(Throwable e) {
-
 			/*[IF Sidecar19-SE]*/
 			if (e instanceof Error) {
 				throw e;
 			}
 			/*[ENDIF]*/
 
-			if (type == null) {
-				throw new BootstrapMethodError(e);
-			}
-			
-			/* create an exceptionHandle with appropriate drop adapter and install that */
-			try {
-				MethodHandle thrower = MethodHandles.throwException(type.returnType(), BootstrapMethodError.class);
-				MethodHandle constructor = MethodHandles.Lookup.IMPL_LOOKUP.findConstructor(BootstrapMethodError.class, MethodType.methodType(void.class, Throwable.class));
-				result = MethodHandles.foldArguments(thrower, constructor.bindTo(e));
-				result = MethodHandles.dropArguments(result, 0, type.parameterList()); 
-			} catch (IllegalAccessException iae) {
-				throw new Error(iae);
-			} catch (NoSuchMethodException nsme) {
-				throw new Error(nsme);
+			if (indyNotCondy) { /* invoke dynamic */
+				if (type == null) {
+					throw new BootstrapMethodError(e);
+				}
+				
+				/* create an exceptionHandle with appropriate drop adapter and install that */
+				try {
+					MethodHandle thrower = MethodHandles.throwException(((MethodType)type).returnType(), BootstrapMethodError.class);
```

Comment:
Hey @DanHeidinga Jack and I were talking through this error handling section but neither of us were too sure about the purpose of these few lines which ends up returning a MethodHandle that will throw a BootstrapMethodError instead of just throwing the error outright like in constant dynamic. Can you help us understand why this is done the way it is?

Reply:
We need to record the failure so that the same error is thrown everytime.  If we didn't record the error, we'd have to rerun the bootstrap method the next time the invokedynamic instruction was executed.  The spec says we need to rethrow the same error